File: MannDrops/ExerciseDifficulty.swift
//
//  ExerciseDifficulty.swift
//  MannDrops
//
//  Created by hackeru on 10/10/16.
//  Copyright © 2016 hackeru. All rights reserved.
//

import UIKit

let typesTag = "OperatorTypes"
let maxNumberTag = "MaxNumber"
var info : [String : Any] = [:]

enum ExerciseDifficulty{
    case EASY//([ExerciseFactory],Int)
    case NORMAL//([ExerciseFactory],Int)
    case HARD//([ExerciseFactory],Int)
    case CRAZY//([ExerciseFactory],Int)
    
    func generateDifficulatyInfo(){
       
        switch self {
            
        case .EASY:
            info[typesTag] = [OperatorType.Plus, OperatorType.Minus]
            info[maxNumberTag] = 10
            
        case .NORMAL:
            info[typesTag] = [OperatorType.Plus, OperatorType.Minus, OperatorType.Multiply, OperatorType.Divide]
            info[maxNumberTag] = 10
            
        case .HARD:
            info[typesTag] = [OperatorType.Plus, OperatorType.Minus, OperatorType.Multiply, OperatorType.Divide]
            info[maxNumberTag] = 15
            
        default: break
            
        }
    }
    
    func generateRandomOperator(info : [String : Any]) -> OperatorType{
        let types : [Any] = info[typesTag] as! [Any]
        let randomIndex = Int(arc4random_uniform(UInt32(types.count)))
        return types[randomIndex] as! OperatorType
    }
    
    func generateExcercise(model : ExercisesDataModel) -> String{
        generateDifficulatyInfo()
        let factory =  ExerciseFactory()
        return factory.generateExercise(maxNumber: info[maxNumberTag] as! Int)
    }
    /*
     func generateExercise(factories : [ExerciseFactory], maxNumber : Int) -> String{
     let randFactoryNumber = Int(arc4random_uniform(UInt32(factories.count - 1)))
     let factory = factories[randFactoryNumber]
     return factory.generateExercise(maxNumber: maxNumber)
     }*/
}

File: java/ExerciseDifficulty.java
import java.util.Arrays;
import java.util.List;

/**
 * Created by dekel31 on 10/8/2016.
 */
public enum ExerciseDifficulty {
    EASY(Arrays.asList(
            new ExerciseFactory(Arrays.asList(Operator.MINUS, Operator.PLUS), 10))),
    NORMAL(Arrays.asList(
            new ExerciseFactory(Arrays.asList(Operator.MINUS, Operator.PLUS), 30),
            new ExerciseFactory(Arrays.asList(Operator.DIVIDE, Operator.MULTIPLY), 10)));
//    HARD(Arrays.asList(new ExerciseFactory(Arrays.asList(Operator.MINUS, Operator.PLUS), 45))),
//    CRAZY(Arrays.asList(new ExerciseFactory(Arrays.asList(Operator.MINUS, Operator.PLUS), 45)));

    private final List<ExerciseFactory> factories;

    ExerciseDifficulty(List<ExerciseFactory> factories) {
        this.factories = factories;
    }

    public String generateExercise(){
        //random between all factories
        long random = Math.round(Math.random() * factories.size());
        return factories.get(random).newInstance();
    }

    /*
    *
    * every instance of enum has a list of factories.
    * when generating an exercise we randomally choose
    * a factory from the enum's list (may include add and substruct or add, substruct, multiply and divide).
    *
    * we then generate the emoji based on the enum value, I.E:
    * emoji gets the random operator and the hard coded enum's max number and then generates it self
    * with a new math exercise.
    *
    *
    * generating two new numbers and embedding the with the operator
    * will be done in the EXERCISE FACTORY.
    * */
}